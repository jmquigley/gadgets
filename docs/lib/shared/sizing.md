<a name="defaultSize"></a>

## defaultSize
Create a single instance with the default values.  Don't instantiate this class
over and over.

**Kind**: global variable  
