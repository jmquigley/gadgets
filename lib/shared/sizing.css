$baseFontSize: 16;

$xxsmall: 6px;
$xxsmall-fs: 0.375em;

$xsmall: 8px;
$xsmall-fs: 0.5em;

$small: 12px;
$small-fs: 0.75em;

$medium: 16px;
$medium-fs: 1.0em;
$normal: $medium;
$normal-fs: $medium-fs;

$large: 24px;
$large-fs: 1.5em;

$xlarge: 32px;
$xlarge-fs: 2.0em;

$xxlarge: 48px;
$xxlarge-fs: 3.0em;

:local {

	.xxsmall {
		font-size: $xxsmall-fs;
		font-weight: inherit;
	}

	.xsmall {
		font-size: $xsmall-fs;
		font-weight: inherit;
	}

	.small {
		font-size: $small-fs;
		font-weight: inherit;
	}

	.medium, .normal {
		font-size: $medium-fs;
		font-weight: inherit;
	}

	.large {
		font-size: $large-fs;
		font-weight: inherit;
	}

	.xlarge {
		font-size: $xlarge-fs;
		font-weight: inherit;
	}

	.xxlarge {
		font-size: $xxlarge-fs;
		font-weight: inherit;
	}

	.xxsmallBorder {
		border: solid 1px;
		/* padding: 1px; */
	}

	.xsmallBorder {
		border: solid 1px;
		/* padding: 2px; */
	}

	.smallBorder {
		border: solid 2px;
		/* padding: 4px; */
	}

	.mediumBorder, .normalBorder {
		border: solid 3px;
		/* padding: 6px; */
	}

	.largeBorder {
		border: solid 3px;
		/* padding: 7px; */
	}

	.xlargeBorder {
		border: solid 3px;
		/* padding: 8px; */
	}

	.xxlargeBorder {
		border: solid 4px;
		/* padding: 10px; */
	}

	.xxsmallBox {
		width: $xxsmall;
		height: $xxsmall;
	}

	.xsmallBox {
		width: $xsmall;
		height: $xsmall;
	}

	.smallBox {
		width: $small;
		height: $small;
	}

	.mediumBox, .normalBox {
		width: $medium;
		height: $medium;
	}

	.largeBox {
		width: $large;
		height: $large;
	}

	.xlargeBox {
		width: $xlarge;
		height: $xlarge;
	}

	.xxlargeBox {
		width: $xxlarge;
		height: $xxlarge;
	}

}
