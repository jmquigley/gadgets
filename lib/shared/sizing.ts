/*
 * Handles management of sizes for components in the library.  This includes:
 *
 * - Fonts (.fontStyle)
 * - Button boxes (.boxStyle)
 * - Borders (.borderStyle)
 *
 * This class is a singleton that computes a range of sizes based on a
 * requested base size (normal).  There are 7 basic sizings in the library:
 *
 * - xxsmall (-10px from normal)
 * - xsmall (-8px from normal)
 * - small (-4px from normal)
 * - normal/medium (base font size)
 * - large (+8px from normal)
 * - xlarge (+16px from normal)
 * - xxlarge (+32px from normal)
 *
 * When a component is created in this library it will have a Sizing
 * parameter (part of the BaseProps).  This sizing parameter determines the
 * font and box sizes for all components.
 *
 * #### Examples
 *
 * ```javascript
 * import {Sizes} from 'gadgets';
 *
 * // Creates a default object with a 16px base font size.
 * const sizes = Sizes.instance(16);
 * ```
 *
 * This would retrieve an instance of a sizing object based on a `normal` size
 * of 16px (1.0em).  Each of the style class strings could then be retrieved
 * and used within a component.  With the current design this is handled
 * automatically within the `BaseComponent` class; it will create and expose
 * this object (composition).
 *
 * Each of the sizes are offsets from normal/medium, where normal uses
 * the typical constant of 16px = 1.0em.
 *
 * The library uses these settings to assign sizes so the UI components can
 * be scaled/responsive to size changes.
 *
 */

'use strict';

const debug = require('debug')('sizing');

import {css} from 'styled-components';

const instances = new Map();

export enum Sizing {
	xxsmall = 'xxsmall',
	xsmall = 'xsmall',
	small = 'small',
	normal = 'normal',
	medium = 'medium',
	large = 'large',
	xlarge = 'xlarge',
	xxlarge = 'xxlarge',
	inherit = 'inherit'
}

export interface FontStyle {
	size: number;
	sizerem: string;
	sizepx: string;
	style: string;
}

export interface Styling {
	type: Sizing;
	borderStyle: string;
	boxStyle: string;
	font: FontStyle;
	rectStyle: string;
}

export const borderStyle: any = {
	[Sizing.xxsmall]: css`border: solid 0.1em;`,
	[Sizing.xsmall]: css`border: solid 0.1em;`,
	[Sizing.small]: css`border: solid 0.125em;`,
	[Sizing.medium]: css`border: solid 0.15em;`,
	[Sizing.normal]: css`border: solid 0.15em;`,
	[Sizing.large]: css`border: solid 0.2em;`,
	[Sizing.xlarge]: css`border: solid 0.25em;`,
	[Sizing.xxlarge]: css`border: solid 0.3em;`,
	[Sizing.inherit]: css`border: inherit;`
};
export const boxStyle: any = {};
export const fontStyle: any = {};

export class Sizes {

	public static readonly styles = require('./sizing.css');
	private _sizes: any = {};

	public static instance(baseFontSize: number): Sizes {
		let tmp = instances.get(baseFontSize);
		if (!tmp) {
			tmp = new Sizes(baseFontSize);
			instances.set(baseFontSize, tmp);
		}

		return tmp._sizes;
	}

	private constructor(baseFontSize: number) {

		const sizes = [
			[Sizing.xxsmall, -10],
			[Sizing.xsmall, -8],
			[Sizing.small, -4],
			[Sizing.medium, 0],
			[Sizing.normal, 0],
			[Sizing.large, 8],
			[Sizing.xlarge, 16],
			[Sizing.xxlarge, 32],
			[Sizing.inherit, 0]
		];

		for (const [key, val] of sizes) {
			const valSize = baseFontSize + Number(val);

			boxStyle[key] = css`
				width: ${valSize}px;
				height: ${valSize}px;
			`;

			fontStyle[key] = css`
				font-size: ${valSize / baseFontSize}rem;
				font-weight: inherit;
			`;

			this._sizes[key] = {
				type: key,
				borderStyle: Sizes.styles[`${key}Border`],
				boxStyle: Sizes.styles[`${key}Box`],
				rectStyle: Sizes.styles[`${key}Rect`],
				font: {
					size: valSize,
					sizerem: `${valSize / baseFontSize}rem`,
					sizepx: `${valSize}px`,
					style: Sizes.styles[key]
				}
			};
		}

		debug('Sizing: %O, boxStyle: %O, fontStyle: %O',
			this._sizes,
			boxStyle,
			fontStyle
		);
	}

	public toString(): string {
		return JSON.stringify(this._sizes, null, 4);
	}

}
