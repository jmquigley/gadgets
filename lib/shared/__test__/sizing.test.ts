'use strict';

// TODO: redo Sizing test module.  Jest and CSS modules break most of this

import * as assert from 'assert';
import {Sizes, Sizing} from '../sizing';

test('Testing creation of the Sizing object', () => {
	const sizes = new Sizes();
	assert(sizes);
	assert(typeof sizes.toString() === 'string');
});

test('Testing contents of FontSize xxsmall', () => {
	const sizes = new Sizes();

	assert(sizes);
	sizes.currentSizing = Sizing.xxsmall;

	assert(sizes.type === Sizing.xxsmall);
	assert(sizes.currentSizing === Sizing.xxsmall);
	assert(sizes.font.sizeem === '0.375em');
	assert(sizes.font.sizepx === '6px');

	assert(sizes.fontSize === 6);
	assert(sizes.font.size === 6);

	assert(sizes.fontStyle === Sizing.xxsmall);
	assert(sizes.font.style === Sizing.xxsmall);

	assert(sizes.borderStyle === 'xxsmallBorder');
	assert(sizes.boxStyle === 'xxsmallBox');
	assert(sizes.next.type === Sizing.xsmall);
	assert(sizes.prev.type === Sizing.xxsmall);

	assert(sizes.getSizing(Sizing.xxsmall).type === Sizing.xxsmall);
});

test('Testing contents of FontSize xsmall', () => {
	const sizes = new Sizes();

	assert(sizes);
	sizes.currentSizing = Sizing.xsmall;

	assert(sizes.type === Sizing.xsmall);
	assert(sizes.currentSizing === Sizing.xsmall);
	assert(sizes.font.sizeem === '0.5em');
	assert(sizes.font.sizepx === '8px');

	assert(sizes.fontSize === 8);
	assert(sizes.font.size === 8);

	assert(sizes.fontStyle === Sizing.xsmall);
	assert(sizes.font.style === Sizing.xsmall);

	assert(sizes.borderStyle === 'xsmallBorder');
	assert(sizes.boxStyle === 'xsmallBox');
	assert(sizes.next.type === Sizing.small);
	assert(sizes.prev.type === Sizing.xxsmall);

	assert(sizes.getSizing(Sizing.xsmall).type === Sizing.xsmall);
});

test('Testing contents of FontSize small', () => {
	const sizes = new Sizes();

	assert(sizes);
	sizes.currentSizing = Sizing.small;

	assert(sizes.type === Sizing.small);
	assert(sizes.currentSizing === Sizing.small);
	assert(sizes.font.sizeem === '0.75em');
	assert(sizes.font.sizepx === '12px');

	assert(sizes.fontSize === 12);
	assert(sizes.font.size === 12);

	assert(sizes.fontStyle === Sizing.small);
	assert(sizes.font.style === Sizing.small);

	assert(sizes.borderStyle === 'smallBorder');
	assert(sizes.boxStyle === 'smallBox');
	assert(sizes.next.type === Sizing.normal);
	assert(sizes.prev.type === Sizing.xsmall);

	assert(sizes.getSizing(Sizing.small).type === Sizing.small);
});

test('Testing contents of FontSize normal', () => {
	const sizes = new Sizes();

	assert(sizes);

	assert(sizes.type === Sizing.normal);
	assert(sizes.currentSizing === Sizing.normal);
	assert(sizes.font.sizeem === '1em');
	assert(sizes.font.sizepx === '16px');

	assert(sizes.fontSize === 16);
	assert(sizes.font.size === 16);

	assert(sizes.fontStyle === Sizing.normal);
	assert(sizes.font.style === Sizing.normal);

	assert(sizes.borderStyle === 'normalBorder');
	assert(sizes.boxStyle === 'normalBox');
	assert(sizes.next.type === Sizing.large);
	assert(sizes.prev.type === Sizing.small);

	assert(sizes.getSizing(Sizing.normal).type === Sizing.normal);
});

test('Testing contents of FontSize large', () => {
	const sizes = new Sizes();

	assert(sizes);
	sizes.currentSizing = Sizing.large;

	assert(sizes.type === Sizing.large);
	assert(sizes.currentSizing === Sizing.large);
	assert(sizes.font.sizeem === '1.5em');
	assert(sizes.font.sizepx === '24px');

	assert(sizes.fontSize === 24);
	assert(sizes.font.size === 24);

	assert(sizes.fontStyle === Sizing.large);
	assert(sizes.font.style === Sizing.large);

	assert(sizes.borderStyle === 'largeBorder');
	assert(sizes.boxStyle === 'largeBox');
	assert(sizes.next.type === Sizing.xlarge);
	assert(sizes.prev.type === Sizing.normal);

	assert(sizes.getSizing(Sizing.large).type === Sizing.large);
});

test('Testing contents of FontSize xlarge', () => {
	const sizes = new Sizes();

	assert(sizes);
	sizes.currentSizing = Sizing.xlarge;

	assert(sizes.type === Sizing.xlarge);
	assert(sizes.currentSizing === Sizing.xlarge);
	assert(sizes.font.sizeem === '2em');
	assert(sizes.font.sizepx === '32px');

	assert(sizes.fontSize === 32);
	assert(sizes.font.size === 32);

	assert(sizes.fontStyle === Sizing.xlarge);
	assert(sizes.font.style === Sizing.xlarge);

	assert(sizes.borderStyle === 'xlargeBorder');
	assert(sizes.boxStyle === 'xlargeBox');
	assert(sizes.next.type === Sizing.xxlarge);
	assert(sizes.prev.type === Sizing.large);

	assert(sizes.getSizing(Sizing.xlarge).type === Sizing.xlarge);
});

test('Testing contents of FontSize xxlarge', () => {
	const sizes = new Sizes();

	assert(sizes);
	sizes.currentSizing = Sizing.xxlarge;

	assert(sizes.type === Sizing.xxlarge);
	assert(sizes.currentSizing === Sizing.xxlarge);
	assert(sizes.font.sizeem === '3em');
	assert(sizes.font.sizepx === '48px');

	assert(sizes.fontSize === 48);
	assert(sizes.font.size === 48);

	assert(sizes.fontStyle === Sizing.xxlarge);
	assert(sizes.font.style === Sizing.xxlarge);

	assert(sizes.borderStyle === 'xxlargeBorder');
	assert(sizes.boxStyle === 'xxlargeBox');
	assert(sizes.next.type === Sizing.xxlarge);
	assert(sizes.prev.type === Sizing.xlarge);

	assert(sizes.getSizing(Sizing.xxlarge).type === Sizing.xxlarge);
});

// TODO: create a sizing object with different base
