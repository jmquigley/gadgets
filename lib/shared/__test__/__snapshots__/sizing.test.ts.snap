// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing creation of the Sizing object with default 1`] = `
Object {
  "inherit": Object {
    "borderStyle": "inheritBorder",
    "boxStyle": "inheritBox",
    "font": Object {
      "size": 16,
      "sizepx": "16px",
      "sizerem": "1rem",
      "style": "inherit",
    },
    "type": "inherit",
  },
  "large": Object {
    "borderStyle": "largeBorder",
    "boxStyle": "largeBox",
    "font": Object {
      "size": 24,
      "sizepx": "24px",
      "sizerem": "1.5rem",
      "style": "large",
    },
    "type": "large",
  },
  "medium": Object {
    "borderStyle": "mediumBorder",
    "boxStyle": "mediumBox",
    "font": Object {
      "size": 16,
      "sizepx": "16px",
      "sizerem": "1rem",
      "style": "medium",
    },
    "type": "medium",
  },
  "normal": Object {
    "borderStyle": "normalBorder",
    "boxStyle": "normalBox",
    "font": Object {
      "size": 16,
      "sizepx": "16px",
      "sizerem": "1rem",
      "style": "normal",
    },
    "type": "normal",
  },
  "small": Object {
    "borderStyle": "smallBorder",
    "boxStyle": "smallBox",
    "font": Object {
      "size": 12,
      "sizepx": "12px",
      "sizerem": "0.75rem",
      "style": "small",
    },
    "type": "small",
  },
  "xlarge": Object {
    "borderStyle": "xlargeBorder",
    "boxStyle": "xlargeBox",
    "font": Object {
      "size": 32,
      "sizepx": "32px",
      "sizerem": "2rem",
      "style": "xlarge",
    },
    "type": "xlarge",
  },
  "xsmall": Object {
    "borderStyle": "xsmallBorder",
    "boxStyle": "xsmallBox",
    "font": Object {
      "size": 8,
      "sizepx": "8px",
      "sizerem": "0.5rem",
      "style": "xsmall",
    },
    "type": "xsmall",
  },
  "xxlarge": Object {
    "borderStyle": "xxlargeBorder",
    "boxStyle": "xxlargeBox",
    "font": Object {
      "size": 48,
      "sizepx": "48px",
      "sizerem": "3rem",
      "style": "xxlarge",
    },
    "type": "xxlarge",
  },
  "xxsmall": Object {
    "borderStyle": "xxsmallBorder",
    "boxStyle": "xxsmallBox",
    "font": Object {
      "size": 6,
      "sizepx": "6px",
      "sizerem": "0.375rem",
      "style": "xxsmall",
    },
    "type": "xxsmall",
  },
}
`;

exports[`Testing creation of the Sizing object with size 24 1`] = `
Object {
  "inherit": Object {
    "borderStyle": "inheritBorder",
    "boxStyle": "inheritBox",
    "font": Object {
      "size": 24,
      "sizepx": "24px",
      "sizerem": "1rem",
      "style": "inherit",
    },
    "type": "inherit",
  },
  "large": Object {
    "borderStyle": "largeBorder",
    "boxStyle": "largeBox",
    "font": Object {
      "size": 32,
      "sizepx": "32px",
      "sizerem": "1.3333333333333333rem",
      "style": "large",
    },
    "type": "large",
  },
  "medium": Object {
    "borderStyle": "mediumBorder",
    "boxStyle": "mediumBox",
    "font": Object {
      "size": 24,
      "sizepx": "24px",
      "sizerem": "1rem",
      "style": "medium",
    },
    "type": "medium",
  },
  "normal": Object {
    "borderStyle": "normalBorder",
    "boxStyle": "normalBox",
    "font": Object {
      "size": 24,
      "sizepx": "24px",
      "sizerem": "1rem",
      "style": "normal",
    },
    "type": "normal",
  },
  "small": Object {
    "borderStyle": "smallBorder",
    "boxStyle": "smallBox",
    "font": Object {
      "size": 20,
      "sizepx": "20px",
      "sizerem": "0.8333333333333334rem",
      "style": "small",
    },
    "type": "small",
  },
  "xlarge": Object {
    "borderStyle": "xlargeBorder",
    "boxStyle": "xlargeBox",
    "font": Object {
      "size": 40,
      "sizepx": "40px",
      "sizerem": "1.6666666666666667rem",
      "style": "xlarge",
    },
    "type": "xlarge",
  },
  "xsmall": Object {
    "borderStyle": "xsmallBorder",
    "boxStyle": "xsmallBox",
    "font": Object {
      "size": 16,
      "sizepx": "16px",
      "sizerem": "0.6666666666666666rem",
      "style": "xsmall",
    },
    "type": "xsmall",
  },
  "xxlarge": Object {
    "borderStyle": "xxlargeBorder",
    "boxStyle": "xxlargeBox",
    "font": Object {
      "size": 56,
      "sizepx": "56px",
      "sizerem": "2.3333333333333335rem",
      "style": "xxlarge",
    },
    "type": "xxlarge",
  },
  "xxsmall": Object {
    "borderStyle": "xxsmallBorder",
    "boxStyle": "xxsmallBox",
    "font": Object {
      "size": 14,
      "sizepx": "14px",
      "sizerem": "0.5833333333333334rem",
      "style": "xxsmall",
    },
    "type": "xxsmall",
  },
}
`;
